### PR TITLE
support UPDATE and DELETE in new plan

### PR DESCRIPTION
**Pull request by XuHuaiyu**
@shenli @hanfei1991 @coocood @zimulala @tiancaiamao PTAL

### CODE REVIEW

File: executor/new_executor_write.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
+package executor
```
Comment (shenli) [inline executor/new_executor_write.go]: Miss license

File: executor/new_builder.go [hunk 2]
```diff
@@ -201,6 +201,10 @@ func (b *executorBuilder) toPBExpr(conditions []expression.Expression, tbl *mode
 
 func (b *executorBuilder) buildSelection(v *plan.Selection) Executor {
 	child := v.GetChildByIndex(0)
+	if v.OldConditions != nil {
```
Comment (hanfei1991) [inline executor/new_builder.go]: set "OldConditions" be a local var.

File: plan/column_pruning.go [hunk 3]
```diff
@@ -358,3 +358,41 @@ func (p *Apply) PruneColumnsAndResolveIndices(parentUsedCols []*expression.Colum
...
+	for i := 0; i < len(orderedList); i++ {
+		if orderedList[i] == nil {
+			continue
+		}
+		orderedList[i].Col.Index = p.GetSchema().GetIndex(orderedList[i].Col)
+		switch assign := orderedList[i].Expr; assign.(type) {
+		case (*expression.Column):
+			assign.(*expression.Column).Index = p.GetSchema().GetIndex(assign.(*expression.Column))
+		case (*expression.ScalarFunction):
+			for i, args := 0, assign.(*expression.ScalarFunction).Args; i < len(args); i++ {
```
Comment (coocood) [inline plan/column_pruning.go]: Column may not be direct argument of the scalar function, like `a = 3 * (a + 1)`

File: executor/new_executor_write.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+		if e.updatedRowKeys[tbl] == nil {
+			e.updatedRowKeys[tbl] = make(map[int64]struct{})
+		}
+		offset := e.getTableOffset(*entry)
+		handle := entry.Handle
+		oldData := row.Data[offset : offset+len(tbl.WritableCols())]
+		newTableData := newData[offset : offset+len(tbl.WritableCols())]
+		_, ok := e.updatedRowKeys[tbl][handle]
+		if ok {
+			// Each matching row is updated once, even if it matches the conditions multiple times.
```
Comment (shenli) [inline executor/new_executor_write.go]: matched row

File: executor/new_executor_write.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	for i := 0; i < len(schema); i++ {
+		s := schema[i]
+		if s.TblName.L == tblName {
+			return i
+		}
+	}
+	return 0
+}
+
+func newUpdateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, updateColumns map[int]*expression.Assignment, t table.Table, offset int, onDuplicateUpdate bool) error {
```
Comment (coocood) [inline executor/new_executor_write.go]: This function is not needed, we can reuse old `updateRecord` method, just change `updateColumns `map[int]*ast.Assignment` to `map[int]bool`, as the value of the map just used to check `nil`, not actually used.

File: executor/new_executor_write.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+		if entry == nil { // outer join
+			continue
+		}
+		tbl := entry.Tbl
+		if e.updatedRowKeys[tbl] == nil {
+			e.updatedRowKeys[tbl] = make(map[int64]struct{})
+		}
+		offset := e.getTableOffset(*entry)
+		handle := entry.Handle
+		oldData := row.Data[offset : offset+len(tbl.WritableCols())]
```
Comment (hanfei1991) [inline executor/new_executor_write.go]: Are you sure the correctness when the columns is pruned.

File: executor/new_executor_write.go [hunk 7]
```diff
@@ -77,11 +75,17 @@ func (e *NewUpdateExec) Next() (*Row, error) {
 		newTableData := newData[offset : offset+len(tbl.WritableCols())]
 		_, ok := e.updatedRowKeys[tbl][handle]
 		if ok {
-			// Each matching row is updated once, even if it matches the conditions multiple times.
+			// Each matched row is updated once, even if it matches the conditions multiple times.
 			continue
 		}
 		// Update row
-		err1 := newUpdateRecord(e.ctx, handle, oldData, newTableData, columns, tbl, offset, false)
+		touched := make(map[int]bool, len(tbl.Cols()))
```
Comment (coocood) [inline executor/new_executor_write.go]: You can change 'getNewUpdateColumns` to return 'touched' instead of 'columns'.

**Comment by coocood** [comment]
LGTM

File: plan/column_pruning.go [hunk 8]
```diff
@@ -358,3 +358,44 @@ func (p *Apply) PruneColumnsAndResolveIndices(parentUsedCols []*expression.Colum
...
+		}
+		orderedList[i].Col.Index = p.GetSchema().GetIndex(orderedList[i].Col)
+		initColumnIndexInExpr(orderedList[i].Expr, p.GetSchema())
+	}
+	p.OrderedList = orderedList
+	return outer, nil
+}
+
+func initColumnIndexInExpr(expr expression.Expression, schema expression.Schema) {
+
```
Comment (zimulala) [inline plan/column_pruning.go]: Remove the empty line.

File: executor/executor_write.go [hunk 9]
```diff
@@ -156,14 +162,12 @@ func (e *UpdateExec) getTableOffset(t table.Table) int {
...
+func updateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, touched map[int]bool, t table.Table, offset int, onDuplicateUpdate bool) error {
 	cols := t.Cols()
-	touched := make(map[int]bool, len(cols))
-
 	assignExists := false
 	var newHandle types.Datum
-	for i, asgn := range updateColumns {
-		if asgn == nil {
+	for i, isTouched := range touched {
+		if !isTouched {
```
Comment (shenli) [inline executor/executor_write.go]: Could this be false?

**Comment by XuHuaiyu** [comment]
I add support to DELETE in this PR, PTAL
@coocood @shenli @zimulala @tiancaiamao @hanfei1991

File: executor/executor_write.go [hunk 10]
```diff
@@ -102,12 +102,16 @@ func (e *UpdateExec) Next() (*Row, error) {
 	return &Row{}, nil
 }
 
-func getUpdateColumns(assignList []*ast.Assignment) (map[int]*ast.Assignment, error) {
-	m := make(map[int]*ast.Assignment, len(assignList))
+func getUpdateColumns(assignList []*ast.Assignment) (map[int]bool, error) {
```
Comment (hanfei1991) [inline executor/executor_write.go]: using slice is enough

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
PTAL @hanfei1991

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**